File: frontend/src/lib/api/gematria.ts
/**
 * Gematria API Client
 *
 * API client for gematria calculations and analysis.
 */
import { apiClient } from './client'

// ============================================================================
// Types
// ============================================================================

export type GematriaSystem =
  | 'hebrew'
  | 'english_ordinal'
  | 'english_reduction'
  | 'transliteration'
  | 'all'

export interface LetterBreakdown {
  letter: string
  value: number
  ordinal_value?: number
  reduced_value?: number
}

export interface GematriaResult {
  value: number
  breakdown: LetterBreakdown[]
  letter_count: number
  original_text: string
  system: string
  final_reduction?: number
  hebrew_text?: string
  meaning?: NumberMeaning | null
  equivalences?: EquivalentWord[]
}

export interface AllSystemsResult {
  original_text: string
  systems: {
    hebrew?: GematriaResult
    english_ordinal?: GematriaResult
    english_reduction?: GematriaResult
    transliteration?: GematriaResult
  }
}

export interface EquivalentWord {
  word: string
  transliteration?: string
  meaning: string
}

export interface EquivalencesResponse {
  value: number
  system: string
  count: number
  words: EquivalentWord[]
}

export interface NumberMeaning {
  name: string
  meaning: string
  keywords: string[]
  hebrew_connection?: string
}

export interface NumberMeaningResponse {
  value: number
  has_known_meaning: boolean
  message?: string
  name?: string
  meaning?: string
  keywords?: string[]
  hebrew_connection?: string
}

export interface AllMeaningsResponse {
  count: number
  meanings: Record<number, NumberMeaning>
}

export interface ProfileGematriaResponse {
  profile_id: string
  profile_name: string
  analysis: AllSystemsResult
}

export interface SystemInfo {
  id: string
  name: string
  description: string
  example: string
}

export interface SystemsResponse {
  systems: SystemInfo[]
}

export interface CalculateRequest {
  text: string
  system?: GematriaSystem
}

// ============================================================================
// API Functions
// ============================================================================

/**
 * Calculate gematria for text
 */
export const calculateGematria = async (
  text: string,
  system: GematriaSystem = 'all'
): Promise<AllSystemsResult | GematriaResult> => {
  const response = await apiClient.post('/gematria/calculate', { text, system })
  return response.data
}

/**
 * Analyze a name comprehensively
 */
export const analyzeName = async (name: string): Promise<AllSystemsResult> => {
  const response = await apiClient.post('/gematria/analyze', { name })
  return response.data
}

/**
 * Find words with the same gematria value
 */
export const getEquivalences = async (
  value: number,
  system: 'hebrew' | 'english' = 'hebrew',
  limit: number = 20
): Promise<EquivalencesResponse> => {
  const response = await apiClient.get(
    `/gematria/equivalences/${value}?system=${system}&limit=${limit}`
  )
  return response.data
}

/**
 * Get the meaning for a specific number
 */
export const getNumberMeaning = async (
  value: number
): Promise<NumberMeaningResponse> => {
  const response = await apiClient.get(`/gematria/meanings/${value}`)
  return response.data
}

/**
 * Get all number meanings
 */
export const getAllMeanings = async (): Promise<AllMeaningsResponse> => {
  const response = await apiClient.get('/gematria/meanings')
  return response.data
}

/**
 * Get gematria analysis for a saved profile
 */
export const getProfileGematria = async (
  profileId: string
): Promise<ProfileGematriaResponse> => {
  const response = await apiClient.get(`/gematria/profile/${profileId}/gematria`)
  return response.data
}

/**
 * Transliterate English text to Hebrew and calculate gematria
 */
export const transliterate = async (text: string): Promise<GematriaResult> => {
  const response = await apiClient.get(
    `/gematria/transliterate?text=${encodeURIComponent(text)}`
  )
  return response.data
}

/**
 * Get information about available gematria systems
 */
export const getSystems = async (): Promise<SystemsResponse> => {
  const response = await apiClient.get('/gematria/systems')
  return response.data
}

// ============================================================================
// Helper Functions
// ============================================================================

/**
 * Get a color class based on the gematria value
 * Uses golden/mystical tones for significant numbers
 */
export const getValueColorClass = (value: number): string => {
  // Special significant numbers get gold/amber tones
  const significantNumbers = [13, 18, 26, 36, 72, 86, 137, 248, 358, 541, 613]
  if (significantNumbers.includes(value)) {
    return 'text-amber-400'
  }

  // Color by digit sum for others
  const digitSum = value.toString().split('').reduce((a, b) => a + parseInt(b), 0) % 9 || 9
  const colors: Record<number, string> = {
    1: 'text-red-400',
    2: 'text-orange-400',
    3: 'text-yellow-400',
    4: 'text-green-400',
    5: 'text-teal-400',
    6: 'text-blue-400',
    7: 'text-indigo-400',
    8: 'text-purple-400',
    9: 'text-pink-400',
  }
  return colors[digitSum] || 'text-gray-400'
}

/**
 * Check if a number has known spiritual significance
 */
export const isSignificantNumber = (value: number): boolean => {
  const significantNumbers = [
    1, 7, 10, 12, 13, 18, 22, 26, 32, 36, 40, 50, 65, 70, 72, 86, 91,
    137, 216, 248, 358, 541, 613
  ]
  return significantNumbers.includes(value)
}

/**
 * Format a Hebrew text for display (RTL)
 */
export const formatHebrewText = (text: string): string => {
  // Just return the text; CSS will handle RTL
  return text
}

/**
 * Get the system display name
 */
export const getSystemDisplayName = (system: string): string => {
  const names: Record<string, string> = {
    hebrew: 'Hebrew',
    english_ordinal: 'English Ordinal',
    english_reduction: 'English Reduction',
    transliteration: 'Transliteration'
  }
  return names[system] || system
}

export default {
  calculateGematria,
  analyzeName,
  getEquivalences,
  getNumberMeaning,
  getAllMeanings,
  getProfileGematria,
  transliterate,
  getSystems,
}

File: backend/app/api/routes/gematria.py
"""
Gematria API Routes

Provides endpoints for gematria calculations across Hebrew and English systems.
"""
from typing import Optional, Literal
from fastapi import APIRouter, HTTPException, Depends
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.services.gematria_service import get_gematria_service
from app.models.birth_data import BirthData

router = APIRouter()


# Request/Response schemas
class CalculateRequest(BaseModel):
    text: str = Field(..., min_length=1, description="Text to calculate gematria for")
    system: Optional[Literal["hebrew", "english_ordinal", "english_reduction", "transliteration", "all"]] = Field(
        default="all",
        description="Which gematria system to use"
    )


class AnalyzeNameRequest(BaseModel):
    name: str = Field(..., min_length=1, description="Name to analyze")


@router.post("/calculate")
async def calculate_gematria(request: CalculateRequest):
    """
    Calculate gematria value for text.

    Supports multiple systems:
    - hebrew: Traditional Hebrew letter values
    - english_ordinal: A=1, B=2...Z=26
    - english_reduction: Each letter reduced to 1-9
    - transliteration: English converted to Hebrew phonetics
    - all: Returns all applicable systems (default)
    """
    service = get_gematria_service()

    if request.system == "all":
        result = service.calculate_all(request.text)
    elif request.system == "hebrew":
        result = service.calculate_hebrew(request.text)
    elif request.system == "english_ordinal":
        result = service.calculate_english_ordinal(request.text)
    elif request.system == "english_reduction":
        result = service.calculate_english_reduction(request.text)
    elif request.system == "transliteration":
        result = service.transliterate(request.text)
    else:
        raise HTTPException(status_code=400, detail=f"Unknown system: {request.system}")

    return result


@router.post("/analyze")
async def analyze_name(request: AnalyzeNameRequest):
    """
    Perform comprehensive gematria analysis on a name.

    Returns gematria values across all systems plus equivalences
    and spiritual meanings for each value.
    """
    service = get_gematria_service()
    return service.analyze_name(request.name)


@router.get("/equivalences/{value}")
async def get_equivalences(
    value: int,
    system: Literal["hebrew", "english"] = "hebrew",
    limit: int = 20
):
    """
    Find words with the same gematria value.

    Args:
        value: The gematria value to search for
        system: Which word database to search (hebrew or english)
        limit: Maximum number of results (default 20)
    """
    if value < 1:
        raise HTTPException(status_code=400, detail="Value must be positive")
    if limit < 1 or limit > 100:
        raise HTTPException(status_code=400, detail="Limit must be between 1 and 100")

    service = get_gematria_service()
    equivalences = service.find_equivalences(value, system, limit)

    return {
        "value": value,
        "system": system,
        "count": len(equivalences),
        "words": equivalences
    }


@router.get("/meanings/{value}")
async def get_number_meaning(value: int):
    """
    Get the spiritual/kabbalistic meaning for a specific number.

    Returns the gematria-specific significance of the number,
    which differs from general numerology meanings.
    """
    if value < 1:
        raise HTTPException(status_code=400, detail="Value must be positive")

    service = get_gematria_service()
    meaning = service.get_number_meaning(value)

    if not meaning:
        return {
            "value": value,
            "has_known_meaning": False,
            "message": f"No specific gematria meaning recorded for {value}"
        }

    return {
        "value": value,
        "has_known_meaning": True,
        **meaning
    }


@router.get("/meanings")
async def get_all_meanings():
    """
    Get all known gematria number meanings.

    Returns meanings for spiritually significant numbers
    in the Kabbalistic tradition.
    """
    service = get_gematria_service()
    meanings = service.get_all_meanings()

    return {
        "count": len(meanings),
        "meanings": meanings
    }


@router.get("/profile/{profile_id}/gematria")
async def get_profile_gematria(profile_id: str, db: Session = Depends(get_db)):
    """
    Calculate gematria analysis for a saved birth profile's name.

    Args:
        profile_id: UUID of the birth data profile
    """
    # Look up the profile
    profile = db.query(BirthData).filter(BirthData.id == profile_id).first()

    if not profile:
        raise HTTPException(status_code=404, detail="Profile not found")

    if not profile.name:
        raise HTTPException(status_code=400, detail="Profile has no name to analyze")

    service = get_gematria_service()
    analysis = service.analyze_name(profile.name)

    return {
        "profile_id": profile_id,
        "profile_name": profile.name,
        "analysis": analysis
    }


@router.get("/transliterate")
async def transliterate_text(text: str):
    """
    Transliterate English text to Hebrew letters and calculate gematria.

    Converts English phonetics to approximate Hebrew letters,
    then calculates the Hebrew gematria value.
    """
    if not text or len(text) < 1:
        raise HTTPException(status_code=400, detail="Text is required")

    service = get_gematria_service()
    return service.transliterate(text)


@router.get("/systems")
async def get_available_systems():
    """
    Get information about available gematria calculation systems.
    """
    return {
        "systems": [
            {
                "id": "hebrew",
                "name": "Hebrew (Standard)",
                "description": "Traditional Kabbalistic values where each Hebrew letter has a fixed numeric value",
                "example": "א=1, ב=2, ג=3...ת=400"
            },
            {
                "id": "english_ordinal",
                "name": "English Ordinal",
                "description": "Simple ordinal values for English letters",
                "example": "A=1, B=2, C=3...Z=26"
            },
            {
                "id": "english_reduction",
                "name": "English Reduction",
                "description": "English values reduced to single digits (1-9)",
                "example": "A=1, J=1, S=1 (all reduce to 1)"
            },
            {
                "id": "transliteration",
                "name": "English to Hebrew",
                "description": "Converts English text to approximate Hebrew phonetics, then calculates Hebrew gematria",
                "example": "David → דויד → 24"
            }
        ]
    }
